### PR TITLE
[spmd_types] FlexAttention: remove hardcoded spmd_types of FlexAttention._complex_flex_attn output

### PR DESCRIPTION
**Pull request by jinsooihm**
`FlexAttention.compiled_flex_attn` runs the compiled flex kernel under `spmd.no_typecheck()` and then re-stamps output's SPMD type. 
It previously hardcoded `PartitionSpec("dp","tp","cp",None)` for `out` (and `"dp","tp","cp"` for `lse`), which bakes in the DP/TP/CP axes and forces CP sharding on the sequence dim regardless of how the layer is actually sharded.

Attention preserves the query's layout (output shape is `(B,N,L,H)` with the same batch/head/seq sharding as `q`), so derive the output type from `q` instead: copy `q`'s local type + shard-dim `PartitionSpec` onto `out`, and the same minus the trailing (unsharded) head-dim entry onto `lse`. 

`spmd_types` has a function `assert_type_like` that can copy local types, without `PartitionSpec`.
But using that would only preserve correctness when `FlexAttention` is used inside a `local_map` region, so it  is not used here.

Edit: spmd_types 0.2.2 `assert_type_like` does have `PartitionSpec`, but this cannot handle the dimension mismatch for `lse`. 

Test plan: existing attention/spmd unit suites pass. Validated e2e under spmd_types + --debug.spmd_typechecking:
  - llama3 debugmodel + flex, context_parallel_degree 2 and 1 (out path)
  - gpt_oss debugmodel + flex (out + lse path via the sink out_transform) all complete step 1 with no SpmdTypeError.

### CODE REVIEW

**Comment by fegin** [comment]
@pianpwk The PR matches what we discussed, but please take a look. We may miss something.

File: torchtitan/models/common/attention.py [hunk 1]
```diff
@@ -20,6 +20,7 @@
 
 import torch
 import torch.nn.functional as F
+from spmd_types.runtime import get_local_type, get_partition_spec
```
Comment (pianpwk) [inline torchtitan/models/common/attention.py]: nit:
```
import spmd_types as spmd
from spmd_types.runtime import get_partition_spec

# later call spmd.get_local_type
```

File: torchtitan/models/common/attention.py [hunk 2]
```diff
@@ -281,9 +287,13 @@ def compiled_flex_attn(
...
             )
         if get_spmd_backend() == "spmd_types" and spmd.is_type_checking():
-            spmd.assert_type(out, spmd.V, spmd.PartitionSpec("dp", "tp", "cp", None))
+            q_local = get_local_type(q)
+            q_ps = get_partition_spec(q)
+            spmd.assert_type(out, q_local, q_ps)
             if return_aux.lse:
-                spmd.assert_type(aux.lse, spmd.V, spmd.PartitionSpec("dp", "tp", "cp"))
+                # lse is (B, N, L) = q minus the trailing (unsharded) head dim.
+                lse_ps = None if q_ps is None else spmd.PartitionSpec(*tuple(q_ps)[:-1])
```
Comment (pianpwk) [inline torchtitan/models/common/attention.py]: nit: don't need the tuple cast

**Review by pianpwk — APPROVED** [review]
thanks!

**Comment by jinsooihm** [comment]
I made the changes. Thank you!

File: torchtitan/models/common/attention.py [hunk 3]
```diff
@@ -264,8 +262,14 @@ def compiled_flex_attn(
...
-        Compiled regions are not currently compatible with SPMD typechecking,
-        so propagate types at the boundary instead of typechecking into Flex.
+        Compiled regions are not currently compatible with SPMD typechecking, so
+        the opaque kernel output is re-typed at the boundary instead of
+        typechecking into Flex. Attention preserves the query's sharding (output
+        is (B, N, L, H) with the same batch/head/seq layout as ``q``), so ``out``
+        takes ``q``'s full SPMD type (local type + shard-dim PartitionSpec), and
+        ``lse`` takes the same minus the trailing (unsharded) head dim. Copying
+        ``q``'s full type keeps this correct even when the module is not wrapped
+        in an outer ``local_map`` boundary that would otherwise restamp ``out``.
```
Comment (fegin) [inline torchtitan/models/common/attention.py]: nit:  Remove `Copying ``q``'s full type keeps this correct ....`. This last statement just refornce the previous `attention preserve the query's sharding ...`

File: torchtitan/models/common/attention.py [hunk 4]
```diff
@@ -281,9 +285,13 @@ def compiled_flex_attn(
                 kernel_options=kernel_options,
             )
         if get_spmd_backend() == "spmd_types" and spmd.is_type_checking():
-            spmd.assert_type(out, spmd.V, spmd.PartitionSpec("dp", "tp", "cp", None))
+            q_local = spmd.get_local_type(q)
+            q_ps = get_partition_spec(q)
```
Comment (fegin) [inline torchtitan/models/common/attention.py:289]: nit: I feel this is a bit asymmetric, directly import `get_partition_spec` while `get_local_type` is accessed through `spmd.get_local_type`. 

cc., @pianpwk Should `get_partition_spec` be exported to the top level just like `get_local_type`.

**Review by fegin — APPROVED** [review]
Thanks for the fix. Please check the comments before merging. You also need to rebase your PR  as it conflicts with your previous one.
 
We should also do the same for VarlenAttention as a followup PR.

File: torchtitan/models/common/attention.py (hunk 4 above)
Comment (pianpwk) [reply torchtitan/models/common/attention.py:289]: spmd types 0.2.3 sir

File: torchtitan/models/common/attention.py [hunk 5]
```diff
@@ -281,9 +285,13 @@ def compiled_flex_attn(
                 kernel_options=kernel_options,
             )
         if get_spmd_backend() == "spmd_types" and spmd.is_type_checking():
-            spmd.assert_type(out, spmd.V, spmd.PartitionSpec("dp", "tp", "cp", None))
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py:284]: > It previously hardcoded PartitionSpec("dp","tp","cp",None) for out (and "dp","tp","cp" for lse), which bakes in the DP/TP/CP axes and forces CP sharding on the sequence dim regardless of how the layer is actually sharded.

From the motivation I wasn't able to tell what's wrong. Without this PR, what error would you hit and when does that happen?

Comment (jinsooihm) [reply torchtitan/models/common/attention.py:284]: There is no error that would happen with current parallelism configuration of the models. 
But it could occur if one wanted to apply CP only on some parts of the model, such as only in the decoder part of a multimodal model.

Comment (tianyu-l) [reply torchtitan/models/common/attention.py:284]: > But it could occur if one wanted to apply CP only on some parts of the model, such as only in the decoder part of a multimodal model.

What error are you getting in this case?

Comment (jinsooihm) [reply torchtitan/models/common/attention.py:284]: When flex attention is used as `inner_attention`, and if you set the `ShardingConfig` of `inner_attention` to be R on CP, it would not propagate correctly because it gets hard coded to be V on CP after the `with spmd.no_typecheck()` block.

Comment (tianyu-l) [reply torchtitan/models/common/attention.py:284]: > it would not propagate correctly because it gets hard coded to be V on CP after the with spmd.no_typecheck() block.

sorry I'm very confused. Which one is correct, why would we set the other thing wrong? And why we don't fix that problem, but instead pursuing this PR?

Comment (jinsooihm) [reply torchtitan/models/common/attention.py:284]: If the user decided to not use CP in the attention for the encoder but use CP in the attention for the decoder, then the sharding configuration for encoder attention could use R on CP.
Then, inside the encoder attention, `q` would come in as R on CP, and `out` should be also R on CP. 
So using `spmd.assert_type(out, spmd.V, spmd.PartitionSpec("dp", "tp", "cp", None))` declares a wrong type for `out` in that use case. 
Without this change, we would need to call `inner_attention` with `no_typecheck` in the `forward` and assert the correct types. 
With this change, we can just set `ShardingConfig` on `inner_attention`.
